Update file.py's initial_diff() to list existing when state: absent

Taoquitok opened the pull request:
##### SUMMARY
Currently when using `file: "path=someFolderPath state=absent"`, if a folder is defined for removing, the diff result only shows the state change, not the files removed. I believe a tweak to `initial_diff` would make sense in the case of using `state=absent` to make it more clear what's to be removed

Fixes #55882

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
[Modules/Files/File.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/file.py)

##### ADDITIONAL INFORMATION
When logging ansible output, either to a file, ARA, or else where, having a detailed record of what was changed is useful for in case anything was later found to be incorrectly updated. This is especially useful when bringing non-ansible managed servers into an ansible stack, as they're likely to be _snowflakes_, with configuration that could break if applying existing roles to it where you would think they overlap.  
  
In my case this issue happened where a default folder that wasn't used any other server got removed from the 1 server it was used on, and lacking a log of what was removed made the recovery process all the more blind.  
As I believe it's generally good practice to run a **--check** run first, adding in a diff of what would be deleted in the situation of the absent statement should protect others from inadvertently removing important files they weren't aware of.
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

###### Example playbook tested with stable 2.7 branch
```yaml
- name: Create example folders
  file:
    path: "{{ item }}"
    state: directory
  with_items:
    - /tmp/example_folder
    - /tmp/example_folder/sub_folder

- name: Create example files
  file:
    path: "{{ item }}"
    state: touch
  with_items:
    - /tmp/example_folder/example_file.tmp
    - /tmp/example_folder/sub_folder/sub_file.tmp

- name: remove example folders and files
  file:
    path: /tmp/example_folder
    state: absent
```

###### Before
```log
TASK [remove example folders and files] ********************************************************************************
--- Output before change
+++ after
@@ -1,13 +1,4 @@
 {
     "path": "/tmp/example_folder",
-    "state": "directory"
+    "state": "absent"
 }
```
###### Output after change
```log
TASK [remove example folders and files] ********************************************************************************
--- before
+++ after
@@ -1,13 +1,4 @@
 {
     "path": "/tmp/example_folder",
-    "path_content": {
-        "directories": [
-            "/tmp/example_folder/sub_folder"
-        ],
-        "files": [
-            "/tmp/example_folder/example_file.tmp",
-            "/tmp/example_folder/sub_folder/sub_file.tmp"
-        ]
-    },
-    "state": "directory"
+    "state": "absent"
 }
```

Taoquitok committed:
Update file.py's initial_diff() to list contents

Taoquitok committed:
Remove trailing whitespace

samdoran commented:
Please create integration tests and a changelog fragment. See [this fragment](https://github.com/ansible/ansible/blob/stable-2.8/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

Taoquitok committed:
WIP integration test

Taoquitok committed:
56353 changelog fragment

Taoquitok committed:
WIP: Removed typo task

Taoquitok committed:
WIP: Force enable diff for test that needs it

Taoquitok committed:
Correct file extension on changelog fragment

Taoquitok committed:
WIP: correcting variable reference

Taoquitok commented:
@samdoran  
Thanks for checking, I've added the change fragment and a basic check that asserts the new diff property is defined.  
How much complexity do you think would be worth adding to confirm the new diff functionality?  
Looking at the current tests there's [a note to organise the tests based on the different states](https://github.com/ansible/ansible/blob/ec8c692b284bf1c91636d62aabafebecf669bcc2/test/integration/targets/file/tasks/main.yml#L42), so if you'd like a bit more complexity on my tests I could start on the `absent` state test file?

samdoran commented on changelogs/fragments/56353-file-extend-absent-diff-result.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - file - Extend `-diff` to return list of files and folders that will be removed in case of `state=absent` (https://github.com/ansible/ansible/pull/56353)
```
This ends up in a restructured text file, so use double backticks for inline code formatting.
```suggestion
  - file - Extend ``-diff`` to return list of files and folders that will be removed in case of ``state=absent`` (https://github.com/ansible/ansible/pull/56353)
```

samdoran commented on test/integration/targets/file/tasks/main.yml:
```diff
@@ -258,6 +258,17 @@
   copy: src=foobar dest={{output_dir}}
 
+- name: check what would be removed if folder state was absent and diff is enabled
+  file: path={{output_dir}} state=absent
```
Please use YAML rather than `k=v` syntax in tests. I know there are a lot of existing tests using `k=v` syntax, but please disregard those and use YAML when writing new tests.
```suggestion
  file: 
    path: "{{ output_dir }}"
    state: absent
```

samdoran commented on test/integration/targets/file/tasks/main.yml:
```diff
@@ -258,6 +258,17 @@
+- name: assert that the absent check lists expected files and folders
+  assert:
+    that:
+      - folder_absent_result.diff.before.path_content is defined
```
I would also loosely verify the contents of the variable such as checking the length of the list or looking for a string in the list contents.

Taoquitok replied:
Thanks, I'll figure out an appropriate test and update soon

Taoquitok replied:
Hey @samdoran,  
I've added a couple of additional asserts to validate that a known folder and file are included in the output. I believe this resolves the conversation, but if not please do re-open it

Taoquitok committed:
Update changelogs/fragments/56353-file-extend-absent-diff-result.yml

Ah, good to know. I'll remember for next time~

Co-Authored-By: Sam Doran <sdoran@redhat.com>

Taoquitok committed:
Update test/integration/targets/file/tasks/main.yml

Ah, I was wondering that, but decided to match the file.  
Thanks for putting in the corrected syntax, will remember for next time.

Co-Authored-By: Sam Doran <sdoran@redhat.com>

Taoquitok committed:
Add additional asserts to validate diff

Taoquitok committed:
Replaced hard coded path root for {{output_dir}}

Taoquitok committed:
Corrected {{output_dir}} to {{ output_dir }}

Taoquitok committed:
test debug to double check variable contents

Taoquitok committed:
trying alternative approach to variable ref

Taoquitok committed:
correcting var: to vars:

Taoquitok committed:
Debug commit to investigate errors

Taoquitok committed:
Cleaned up debug task, and corrected path refs

Updated folder path references to use the folder_absent_result.path param rather than continue to attempt to use the originating output_dir variable. Necessary to get around output_dir being a relative path
